ghcide/src/Development/IDE/Types/Logger.hs

```diff
@@ -27,33 +54,164 @@ data Priority
     | Error -- ^ Such log messages must never occur in expected usage.
     deriving (Eq, Show, Ord, Enum, Bounded)
 
+data LogMessage =
+  LogOther LogOtherMessage
+
+data LogOtherMessage = LogOtherMessage
+  { stack    :: !CallStack
+  , priority :: Priority
+  , payload  :: Text
+  }
```

Comment:
I remember reading this and it wouldn't go into my brain.

How would it look to log something? For instance would it be more verbose like
```haskell
data GhcideLog = GhcideLog Text

logMsg logger $ LogGhcide $ GhcideLog "something is happening"
```
and then how do we get from `LogMessage` to (for instance) the text that is output to `stderr` including the priority? Is it something like this?
```haskell
logMessageToText :: Priority -> LogMessage -> Text

infoLogger = cmap (logMessageToText Info)  textStderrLogger
debugLogger = cmap (logMessageToText Debug) textStderrLogger
```
and then I guess we can recover our `logInfo`, `logDebug` api by doing
```haskell
logInfo text = logMsg_ infoLogger (LogGhcide $ GhcideLog text)
logDebug text = logMsg_ debugLogger (LogGhcide $ GhcideLog text)
-- or more generally
logInfo :: GhcideLog -> IO ()
logInfo msg = logMsg_ infoLogger (LogGhcide msg)
```
but this means that we need to define `logInfo`, `logDebug`, etc per different kind of `LogMessage` if we want the convenience. 

So if `LogMessage` is in the `Main` module then so would the `logMessageToText priority` function, as well as the `infoLogger`, `debugLogger`, ... and so that is what is meant by `Main` deciding things? I guess if we want minimum log levels we'd have conditionals in the functions `infoLogger`, `debugLogger`, etc, to be the empty logger if below the log level.

Or maybe instead each plugin or message type defines its own logger like `Logger GhcideLog` and then Main can convert it into a `Logger LogMessage` and then `Main` eventually passes this logger around to everybody else?

I feel like I could be totally off base.

Reply:
This is a good introduction: https://www.youtube.com/watch?v=qzOQOmmkKEM

Reply:
> how do we get from LogMessage to (for instance) the text that is output to stderr including the priority?

Using `contramap` you transform your `Text` logger to accept more specific log types. 

The priority is subjective and decided by the application, not by the `logMsg` call site. 

To support this in HLS, a few architectural changes will be needed:

1. The `PluginDescriptor` type will need to be parameterised by the type of the plugin log messages
2. The `PluginDescriptor` type will need an additional field to provide a priority function: `logtype -> Priority`
3. The plugins will no longer use the `logger` field from `ShakeExtras`. Instead, they will expect a logger of the type they need as an argument to construct the `PluginDescriptor`

For 2 there are alternative encodings, like wrapping all log types in a `WithDefaultPriority` tuple type.

3 does have some implications - it will make dynamic loading of plugins harder to support if we ever decide to go down that route, which seems very unlikely anyway.

Reply:
> Or maybe instead each plugin or message type defines its own logger like Logger GhcideLog and then Main can convert it into a Logger LogMessage and then Main eventually passes this logger around to everybody else?

This is the way

Reply:
I watched the video and I'm pretty sure I get it now. I'll sleep and then try and do the do.